File: Backend/StrategyEngine/Simulation/Indicators.fs
module Indicators

open AST
open EngineTypes
open System

let private getPriceSeries (ticker: string) (history: FullPriceHistory) : float array option =
    history 
    |> List.tryFind (fun p -> p.Ticker = ticker)
    |> Option.map (fun p -> p.DailyData |> Array.map (fun d -> d.Price))


let private calculateSMA (data: float array) (period: int) (currentDay: int) : float =
    if currentDay < period - 1 then 0.0
    else
        let startIdx = currentDay - period + 1
        let slice = data.[startIdx .. currentDay]
        Array.average slice

let private calculateEMA (data: float array) (period: int) (currentDay: int) : float =
    if currentDay < period - 1 then 0.0
    else
        let warmupPeriod = period * 3
        let startIdx = Math.Max(0, currentDay - warmupPeriod)
        
        let mutable ema = data.[startIdx] 
        
        let k = 2.0 / (float period + 1.0)

        for i in startIdx + 1 .. currentDay do
            ema <- (data.[i] * k) + (ema * (1.0 - k))
        
        ema

let private calculateRSI (data: float array) (period: int) (currentDay: int) : float =
    if currentDay < period then 0.0
    else
        let warmup = period * 3
        let startIdx = Math.Max(1, currentDay - warmup)
        
        let mutable avgGain = 0.0
        let mutable avgLoss = 0.0

        for i in startIdx .. (startIdx + period - 1) do
            let change = data.[i] - data.[i-1]
            if change > 0.0 then avgGain <- avgGain + change
            else avgLoss <- avgLoss + abs(change)
        
        avgGain <- avgGain / float period
        avgLoss <- avgLoss / float period

        for i in (startIdx + period) .. currentDay do
            let change = data.[i] - data.[i-1]
            let gain = if change > 0.0 then change else 0.0
            let loss = if change < 0.0 then abs(change) else 0.0
            
            avgGain <- ((avgGain * float(period - 1)) + gain) / float period
            avgLoss <- ((avgLoss * float(period - 1)) + loss) / float period

        if avgLoss = 0.0 then 100.0
        else
            let rs = avgGain / avgLoss
            100.0 - (100.0 / (1.0 + rs))

let private calculateVol (data: float array) (period: int) (currentDay: int) : float =
    if currentDay < period then 0.0
    else
        let startIdx = currentDay - period + 1
        let returns = 
            [| for i in startIdx .. currentDay do
                 if data.[i-1] <> 0.0 then 
                     yield Math.Log(data.[i] / data.[i-1])
                 else 
                     yield 0.0 |]
        
        let mean = Array.average returns
        let sumSqDiff = returns |> Array.sumBy (fun r -> Math.Pow(r - mean, 2.0))
        let stdDev = Math.Sqrt(sumSqDiff / float (period - 1))
        
        stdDev * Math.Sqrt(252.0)

let private calculateReturn (data: float array) (period: int) (currentDay: int) : float =
    if currentDay < period then 0.0
    else
        let currentPrice = data.[currentDay]
        let pastPrice = data.[currentDay - period]
        if pastPrice = 0.0 then 0.0
        else (currentPrice - pastPrice) / pastPrice

let private calculatePastPrice (data: float array) (period: int) (currentDay: int) : float =
    if currentDay < period then 0.0
    else data.[currentDay - period]


let calculate (indicator: Indicator) (history: FullPriceHistory) (currentDay: int) : float =
    let prices = getPriceSeries indicator.Asset history
    
    match prices with
    | None -> 
        0.0
    | Some data ->
        let p = 
            match indicator.Period with
            | Some v -> v
            | None -> 
                match indicator.IndicatorType with
                | SMA | EMA | Vol -> 20
                | RSI -> 14
                | Return | PastPrice -> 1

        match indicator.IndicatorType with
        | SMA -> calculateSMA data p currentDay
        | EMA -> calculateEMA data p currentDay
        | RSI -> calculateRSI data p currentDay
        | Vol -> calculateVol data p currentDay
        | Return -> calculateReturn data p currentDay
        | PastPrice -> calculatePastPrice data p currentDay

File: Frontend/src/assets/fable_build/Simulation/Indicators.js
import { map } from "../fable_modules/fable-library-js.4.27.0/Option.js";
import { sumBy, item, average, map as map_1 } from "../fable_modules/fable-library-js.4.27.0/Array.js";
import { tryFind } from "../fable_modules/fable-library-js.4.27.0/List.js";
import { max } from "../fable_modules/fable-library-js.4.27.0/Double.js";
import { singleton, collect, delay, toArray } from "../fable_modules/fable-library-js.4.27.0/Seq.js";
import { rangeDouble } from "../fable_modules/fable-library-js.4.27.0/Range.js";

function getPriceSeries(ticker, history) {
    return map((p_1) => map_1((d) => d.Price, p_1.DailyData, Float64Array), tryFind((p) => (p.Ticker === ticker), history));
}

function calculateSMA(data, period, currentDay) {
    if (currentDay < (period - 1)) {
        return 0;
    }
    else {
        const startIdx = ((currentDay - period) + 1) | 0;
        return average(data.slice(startIdx, currentDay + 1), {
            GetZero: () => 0,
            Add: (x_1, y) => (x_1 + y),
            DivideByInt: (x, i) => (x / i),
        });
    }
}

function calculateEMA(data, period, currentDay) {
    if (currentDay < (period - 1)) {
        return 0;
    }
    else {
        const startIdx = max(0, currentDay - (period * 3)) | 0;
        let ema = item(startIdx, data);
        const k = 2 / (period + 1);
        for (let i = startIdx + 1; i <= currentDay; i++) {
            ema = ((item(i, data) * k) + (ema * (1 - k)));
        }
        return ema;
    }
}

function calculateRSI(data, period, currentDay) {
    if (currentDay < period) {
        return 0;
    }
    else {
        const startIdx = max(1, currentDay - (period * 3)) | 0;
        let avgGain = 0;
        let avgLoss = 0;
        for (let i = startIdx; i <= ((startIdx + period) - 1); i++) {
            const change = item(i, data) - item(i - 1, data);
            if (change > 0) {
                avgGain = (avgGain + change);
            }
            else {
                avgLoss = (avgLoss + Math.abs(change));
            }
        }
        avgGain = (avgGain / period);
        avgLoss = (avgLoss / period);
        for (let i_1 = startIdx + period; i_1 <= currentDay; i_1++) {
            const change_1 = item(i_1, data) - item(i_1 - 1, data);
            const gain = (change_1 > 0) ? change_1 : 0;
            const loss = (change_1 < 0) ? Math.abs(change_1) : 0;
            avgGain = (((avgGain * (period - 1)) + gain) / period);
            avgLoss = (((avgLoss * (period - 1)) + loss) / period);
        }
        if (avgLoss === 0) {
            return 100;
        }
        else {
            return 100 - (100 / (1 + (avgGain / avgLoss)));
        }
    }
}

function calculateVol(data, period, currentDay) {
    if (currentDay < period) {
        return 0;
    }
    else {
        const startIdx = ((currentDay - period) + 1) | 0;
        const returns = toArray(delay(() => collect((i) => ((item(i - 1, data) !== 0) ? singleton(Math.log(item(i, data) / item(i - 1, data))) : singleton(0)), rangeDouble(startIdx, 1, currentDay))));
        const mean = average(returns, {
            GetZero: () => 0,
            Add: (x_1, y) => (x_1 + y),
            DivideByInt: (x, i_1) => (x / i_1),
        });
        const sumSqDiff = sumBy((r) => Math.pow(r - mean, 2), returns, {
            GetZero: () => 0,
            Add: (x_2, y_1) => (x_2 + y_1),
        });
        return Math.sqrt(sumSqDiff / (period - 1)) * Math.sqrt(252);
    }
}

function calculateReturn(data, period, currentDay) {
    if (currentDay < period) {
        return 0;
    }
    else {
        const currentPrice = item(currentDay, data);
        const pastPrice = item(currentDay - period, data);
        if (pastPrice === 0) {
            return 0;
        }
        else {
            return (currentPrice - pastPrice) / pastPrice;
        }
    }
}

function calculatePastPrice(data, period, currentDay) {
    if (currentDay < period) {
        return 0;
    }
    else {
        return item(currentDay - period, data);
    }
}

export function calculate(indicator, history, currentDay) {
    const prices = getPriceSeries(indicator.Asset, history);
    if (prices != null) {
        const data = prices;
        let p;
        const matchValue = indicator.Period;
        if (matchValue == null) {
            const matchValue_1 = indicator.IndicatorType;
            switch (matchValue_1.tag) {
                case 2: {
                    p = 14;
                    break;
                }
                case 4:
                case 5: {
                    p = 1;
                    break;
                }
                default:
                    p = 20;
            }
        }
        else {
            p = matchValue;
        }
        const matchValue_2 = indicator.IndicatorType;
        switch (matchValue_2.tag) {
            case 1:
                return calculateEMA(data, p, currentDay);
            case 2:
                return calculateRSI(data, p, currentDay);
            case 3:
                return calculateVol(data, p, currentDay);
            case 4:
                return calculateReturn(data, p, currentDay);
            case 5:
                return calculatePastPrice(data, p, currentDay);
            default:
                return calculateSMA(data, p, currentDay);
        }
    }
    else {
        return 0;
    }
}

